pkg/assembler/backends/ent/backend/vex.go

```diff
@@ -0,0 +1,246 @@
+	funcName := "IngestVEXStatement"
+	if err := helper.ValidatePackageOrArtifactInput(&subject, "IngestVEXStatement"); err != nil {
+		return nil, gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+	if err := helper.ValidateVulnerabilityIngestionInput(vulnerability, "IngestVEXStatement", false); err != nil {
+		return nil, gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+
+	err := validateVexInputBasedOnStatus(vexStatement.Status, vexStatement.VexJustification, vexStatement.Statement)
+	if err != nil {
+		return nil, gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+
+	record, err := WithinTX(ctx, b.client, func(ctx context.Context) (*ent.CertifyVex, error) {
+		client := ent.TxFromContext(ctx)
+		insert := client.CertifyVex.Create()
+
+		conflictColumns := []string{
+			certifyvex.FieldKnownSince,
+			certifyvex.FieldStatus,
+			certifyvex.FieldStatement,
+			certifyvex.FieldStatusNotes,
+			certifyvex.FieldJustification,
+			certifyvex.FieldOrigin,
+			certifyvex.FieldCollector,
+		}
+
+		// manage vulnerability
+		v, err := getAdvisoryFromVulnerabilityInput(ctx, client.Client(), vulnerability)
+		if err != nil {
+			return nil, err
+		}
+		insert.SetVulnerability(v)
+		conflictColumns = append(conflictColumns, certifyvex.FieldVulnerabilityID)
+		var conflictWhere = sql.And(sql.NotNull(certifyvex.FieldVulnerabilityID))
+
+		// manage package or artifact
+		if subject.Package != nil {
+			p, err := getPkgVersion(ctx, client.Client(), *subject.Package)
+			if err != nil {
+				return nil, Errorf("%v ::  %s", funcName, err)
+			}
+			insert.SetPackage(p)
+			conflictColumns = append(conflictColumns, certifyvex.FieldPackageID)
+			conflictWhere = sql.And(
+				sql.NotNull(certifyvex.FieldPackageID),
+				sql.IsNull(certifyvex.FieldArtifactID),
+			)
+		} else if subject.Artifact != nil {
+			art, err := client.Artifact.Query().
+				Where(artifactQueryInputPredicates(*subject.Artifact)).
+				Only(ctx)
+			if err != nil {
+				return nil, Errorf("%v ::  %s", funcName, err)
+			}
+			insert.SetArtifact(art)
+			conflictColumns = append(conflictColumns, certifyvex.FieldArtifactID)
+			conflictWhere = sql.And(
+				sql.IsNull(certifyvex.FieldPackageID),
+				sql.NotNull(certifyvex.FieldArtifactID),
+			)
+		} else {
+			return nil, Errorf("%v :: %s", funcName, "subject must be either a package or artifact")
+		}
+
+		insert.
+			SetKnownSince(vexStatement.KnownSince).
+			SetStatus(vexStatement.Status.String()).
+			SetStatement(vexStatement.Statement).
+			SetStatusNotes(vexStatement.StatusNotes).
+			SetJustification(vexStatement.VexJustification.String()).
+			SetOrigin(vexStatement.Origin).
+			SetCollector(vexStatement.Collector)
+
+		id, err := insert.
+			OnConflict(
+				sql.ConflictColumns(conflictColumns...),
+				sql.ConflictWhere(conflictWhere),
+			).
+			UpdateNewValues().
```

Comment:
For performance reasons it's probably better to use `DoNothing()` here, which won't return anything, but it will massively increase ingestion performance. Since everything is migrating to returning an ID or nothing, we can use the original input spec to query the final result.

Reply:
I've tried to use `DoNothing()` but the `Ingest same twice` VEX test then failed with the `sql: no rows in result set` error.
Investigating this issue, it turned out to be a known issue for Ent (ref. https://github.com/ent/ent/issues/1821) so I've used `Ignore()` which isn't the same but it's the "best" workaround.

Reply:
Ah yes, so instead of calling `.ID()` you'll want `.Exec()` which doesn't expect any results. Using Ignore technically works but it assigns all the existing values to themselves, which is as expensive as doing an update.

Reply:
yep but I found this comment https://github.com/ent/ent/issues/1821#issuecomment-1429948735 reporting the same issue.
But I'll give a try right now and I'll let you know.